### PR TITLE
53 - create html/css for activity history of user item

### PR DESCRIPTION
**Pull request by ghost**
Closes: [Activity History of UserItem](https://vicesoftware.atlassian.net/browse/MAP-53)

Updates:-
1. Added Activity history table.
2. Integrated Manage Subscription modal and Cancel Subscription modal.

### CODE REVIEW

File: webapp/src/app.css [hunk 1]
```diff
@@ -244,8 +263,43 @@ select.form-control,
 .btn-link {
 	color: #294c6a;
 }
+.mb-30 {
+	margin-bottom: 25px !important;
+}
 /*###############-------Btn  css End------###################*/
 
+/*###############-------Modal  css Start------###################*/
```
Comment (himanshuvicesoftware) [inline webapp/src/app.css:270]: we can add the Modal CSS inside the Modal widget

File: webapp/src/app.css [hunk 2]
```diff
@@ -244,8 +263,43 @@ select.form-control,
...
+.modal-footer,
+.modal-header,
+.modal-body {
+	padding: 15px;
+}
+.modal-backdrop {
+	background-color: #0d1e45;
+}
+.upload-container {
```
Comment (himanshuvicesoftware) [inline webapp/src/app.css]: this css for what?

File: webapp/src/features/UserItem/ActivityHistory.js [hunk 3]
```diff
@@ -0,0 +1,232 @@
+import React from 'react'
+import { Row, Col, Button, Image } from 'react-bootstrap'
+import Icons from '../../assets/icons'
+import Table from '../../widgets/Table'
+
+const ActivityHistory = () => {
+	const creditIconFormatter = (row) => {
```
Comment (himanshuvicesoftware) [inline webapp/src/features/UserItem/ActivityHistory.js]: Looks like we haven't use SVG Icons directly and may be we can re-use alot of code there,
Will discuss
@aakashThapa0011

File: webapp/src/features/UserItem/BillingHistory.js [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+import { USERITEM_MODAL_TYPES } from '../UserItemModals/UserItemModals.constants'
+import getUserItemModal from '../UserItemModals'
+
+const BillingHistory = () => {
+	const dispatch = useDispatch()
+	const selectedModal = useSelector(selectedUserItemModal)
+
+	const deleteButtonFormatter = (cell, row) => (
+		<div className='d-inline-flex align-items-center'>
+			{row.revenue === '$115' && (
```
Comment (himanshuvicesoftware) [inline webapp/src/features/UserItem/BillingHistory.js:19]: Why we hardcoded price here

File: webapp/src/features/UserItem/UserInformation.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+		},
+		{
+			name: 'Group',
+			value: 'Brandcave',
+		},
+	]
+	return (
+		<Cards className='mb-25' title='User Information'>
+			<Row className='user-info'>
+				{mockData.map((user, index) => (
```
Comment (himanshuvicesoftware) [inline webapp/src/features/UserItem/UserInformation.js]: index is used for what?

File: webapp/src/features/UserItem/BillingHistory.js [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+			id: 4,
+			invoice: '#123456789',
+			date: '01/01/22',
+			revenue: '$220',
+		},
+		{
+			id: 5,
+			invoice: '#123456789',
+			date: '01/01/22',
+			revenue: '$50',
```
Comment (himanshuvicesoftware) [inline webapp/src/features/UserItem/BillingHistory.js:81]: so we are expecting the price thing with '$' sign or we are going to update this thing with formatter in upcoming PRs

**Review by himanshuvicesoftware — CHANGES_REQUESTED** [review]

File: webapp/src/app.css (hunk 1 above)
Comment (ghost) [reply webapp/src/app.css:270]: We were having an modal size issue when defining the css in Modal widget.

File: webapp/src/app.css (hunk 2 above)
Comment (ghost) [reply webapp/src/app.css]: No need of this css..removed from the code.

File: webapp/src/features/UserItem/ActivityHistory.js (hunk 3 above)
Comment (ghost) [reply webapp/src/features/UserItem/ActivityHistory.js]: updated

File: webapp/src/features/UserItem/BillingHistory.js (hunk 4 above)
Comment (ghost) [reply webapp/src/features/UserItem/BillingHistory.js:19]: We need to update condition later on...for now to show the alert icon we had hardcoded the value.

File: webapp/src/features/UserItem/UserInformation.js (hunk 5 above)
Comment (ghost) [reply webapp/src/features/UserItem/UserInformation.js]: it was not being used, so removed from the code.

File: webapp/src/features/UserItem/BillingHistory.js (hunk 6 above)
Comment (ghost) [reply webapp/src/features/UserItem/BillingHistory.js:81]: Actually revenue value will be received as formatted from the backend due to currency being used can be variable.

File: webapp/src/features/UserItem/ActivityHistory.js [hunk 7]
```diff
@@ -0,0 +1,149 @@
+import React from 'react'
+import { Row, Col, Button, Image } from 'react-bootstrap'
+import Icons from '../../assets/icons'
+import Table from '../../widgets/Table'
+import classNames from 'classnames'
+
+const ActivityHistory = () => {
+	const eventIconFormatter = (row) => {
+		let credit = false
```
Comment (himanshuvicesoftware) [inline webapp/src/features/UserItem/ActivityHistory.js]: Good to have it inside an object.

Comment (himanshuvicesoftware) [reply webapp/src/features/UserItem/ActivityHistory.js]: and used these as keys in object, and can use switch-cases there.

File: webapp/src/Routes.js [hunk 8]
```diff
@@ -2,10 +2,14 @@ import React from 'react'
 import { Switch, Route } from 'react-router-dom'
 import Page from './widgets/Page/Page'
 import Dashboard from './features/Dashboard'
+import UserItem from './features/UserItem'
 
 export default function Routes() {
 	return (
 		<Switch>
+			<PageRoute path='/userItem'>
```
Comment (himanshuvicesoftware) [inline webapp/src/Routes.js]: user-item

File: webapp/src/features/UserItem/ActivityHistory.js [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+			churn = true
+		} else if (row.includes('Upgraded')) {
+			upgraded = true
+		} else if (row.includes('Subscription Creation')) {
+			creation = true
+		}
+		return (
+			<div className='d-flex align-items-center'>
+				<div className='arrow-bubble bubble-sm position-relative rounded-circle overflow-hidden d-flex align-items-center justify-content-center mr-2'>
+					{credit && (
```
Comment (himanshuvicesoftware) [inline webapp/src/features/UserItem/ActivityHistory.js]: can we move this inside a function and use swith case

**Review by himanshuvicesoftware — CHANGES_REQUESTED** [review]

File: webapp/src/Routes.js (hunk 8 above)
Comment (ghost) [reply webapp/src/Routes.js]: updated

File: webapp/src/features/UserItem/ActivityHistory.js (hunk 9 above)
Comment (ghost) [reply webapp/src/features/UserItem/ActivityHistory.js]: updated

File: webapp/src/features/UserItem/ActivityHistory.js [hunk 10]
```diff
@@ -0,0 +1,138 @@
...
+				break
+			default:
+				break
+		}
+		return (
+			<div className='d-flex align-items-center'>
+				<div className='arrow-bubble bubble-sm position-relative rounded-circle overflow-hidden d-flex align-items-center justify-content-center mr-2'>
+					<Image src={Icons[iconName]} alt='events-icon' />
+					<span
+						className={classNames(iconClass, 'position-absolute h-100 w-100')}
```
Comment (himanshuvicesoftware) [inline webapp/src/features/UserItem/ActivityHistory.js:47]: No need of classNames.

**Review by himanshuvicesoftware — CHANGES_REQUESTED** [review]

File: webapp/src/features/UserItem/ActivityHistory.js (hunk 10 above)
Comment (ghost) [reply webapp/src/features/UserItem/ActivityHistory.js:47]: classNames need to be provided cause iconClass is the className which changes according to the data....I have tried using `` formatting but it does not support in className.

File: webapp/src/Routes.js [hunk 11]
```diff
@@ -2,10 +2,14 @@ import React from 'react'
 import { Switch, Route } from 'react-router-dom'
 import Page from './widgets/Page/Page'
```
Comment (sahilvicesoftware) [inline webapp/src/Routes.js:3]: Change the title of the app in file - `/webapp/public/index.html` from 
React App to `MapRight Revenue Dashboard`

File: webapp/src/Routes.js [hunk 12]
```diff
@@ -2,10 +2,14 @@ import React from 'react'
...
 import Dashboard from './features/Dashboard'
+import UserItem from './features/UserItem'
 
 export default function Routes() {
 	return (
 		<Switch>
+			<PageRoute path='/user-Item'>
```
Comment (sahilvicesoftware) [inline webapp/src/Routes.js]: /user-Item - we need to load the same user's details when we reload the page
for that we need to know the userId as well
so the page url should be `users/{userId}`
where on reload we would fetch the userId from params and show the details

File: webapp/src/app.css [hunk 13]
```diff
@@ -1,6 +1,12 @@
+.bg-light {
+	background-color: #f7f7fc !important;
```
Comment (sahilvicesoftware) [inline webapp/src/app.css:2]: make sure these color codes are from the UX Specs file

File: webapp/src/app.css [hunk 14]
```diff
@@ -230,8 +249,8 @@ select.form-control,
 	border-color: #273764;
 }
 .btn-outline-primary:hover {
-	color: #fff;
-	background-color: #273764;
+	color: #273765;
+	background-color: #2737650f;
```
Comment (sahilvicesoftware) [inline webapp/src/app.css:253]: this color code looks wrong to me

File: webapp/src/widgets/NavBar/NavBar.js [hunk 15]
```diff
@@ -1,10 +1,18 @@
-import React from 'react'
+import React, { useState, useEffect } from 'react'
 import { Navbar, Nav, Container, Image } from 'react-bootstrap'
 import { LinkContainer } from 'react-router-bootstrap'
 import Icons from '../../assets/icons'
```
Comment (sahilvicesoftware) [inline webapp/src/widgets/NavBar/NavBar.js:4]: use index file for importing images as well

File: webapp/src/widgets/NavBar/NavBar.js [hunk 16]
```diff
@@ -16,7 +24,7 @@ export default function NavBar() {
 				<Navbar.Toggle aria-controls='basic-navbar-nav' />
 				<Navbar.Collapse className='collapse'>
 					<Nav className='mr-auto align-items-lg-center'>
-						<Nav.Item className='active'>
+						<Nav.Item className={classNames({ active: selectedNav === '' })}>
```
Comment (sahilvicesoftware) [inline webapp/src/widgets/NavBar/NavBar.js]: `selectedNav === '' ` ?
is this in progress?

File: webapp/src/widgets/Modal/Modal.js [hunk 17]
```diff
@@ -3,7 +3,6 @@ import { Modal, Image } from 'react-bootstrap'
 import { useSelector, useDispatch } from 'react-redux'
 import { selectShowModal } from './modal.selectors'
 import { actions } from './modal.slice'
-import './Modal.css'
```
Comment (sahilvicesoftware) [inline webapp/src/widgets/Modal/Modal.js:6]: why is this removed?

File: webapp/src/features/Dashboard/SubscriptionTable.js [hunk 18]
```diff
@@ -13,8 +13,11 @@ import Table from '../../widgets/Table'
...
 import Icons from '../../assets/icons'
 import classNames from 'classnames'
+import { useHistory } from 'react-router-dom'
+import SortingCustomIcon from '../../widgets/SortingCustomIcon'
 
 const SubscriptionTable = () => {
+	const history = useHistory()
 	const viewButton = (cell, row) => (
 		<div className='d-inline-flex align-items-center'>
 			{row.userName === 'Test1' ? (
```
Comment (sahilvicesoftware) [inline webapp/src/features/Dashboard/SubscriptionTable.js:23]: add the TODOs where we need to remove the mock data

File: webapp/src/widgets/SortingCustomIcon/SortingCustomIcon.js [hunk 19]
```diff
@@ -0,0 +1,32 @@
...
+					<Image
+						src={row.userColumn ? Icons.caretDarkIcon : Icons.caretIcon}
+						alt='caret-icon'
+						className='opacity-50 upArrow'
+						width='10'
+					/>
+				</span>
+			)}
+			{(order === 'desc' || !order) && (
+				<span className='arrow-down'>
```
Comment (sahilvicesoftware) [inline webapp/src/widgets/SortingCustomIcon/SortingCustomIcon.js:19]: i think only 2 classes are different
we can use npm classNames to add conditional classes rather than separate `<span>`

File: webapp/src/widgets/Cards/Cards.js [hunk 20]
```diff
@@ -1,20 +1,53 @@
 import React from 'react'
-import { Card } from 'react-bootstrap'
+import { Card, Button } from 'react-bootstrap'
 import classNames from 'classnames'
 import './Cards.css'
+import { USERITEM_MODAL_TYPES } from '../../features/UserItemModals/UserItemModals.constants'
+import { useDispatch, useSelector } from 'react-redux'
+import { setSelectedUserItemModal } from '../../features/UserItem/userItem.slice'
+import { selectedUserItemModal } from '../../features/UserItem/userItem.selectors'
```
Comment (sahilvicesoftware) [inline webapp/src/widgets/Cards/Cards.js]: use `select` as prefix in selectors not `selected. so, selectUserItemModal

File: webapp/src/widgets/Cards/Cards.js [hunk 21]
```diff
@@ -1,20 +1,53 @@
...
+		<>
+			{selectedModal && getUserItemModal(selectedModal)}
+			<Card
+				className={classNames('shadow-sm border-0 mb-25', {
+					'h-100': props.title === 'Filter',
+				})}
+			>
+				<Card.Body className='p-15'>
+					{props.title === 'Billing History' ? (
+						<div className='d-flex align-items-center'>
```
Comment (sahilvicesoftware) [inline webapp/src/widgets/Cards/Cards.js]: need to discuss this

File: webapp/src/features/UserItemModals/AddInvoiceModal.js [hunk 22]
```diff
@@ -0,0 +1,101 @@
...
+				<Col md={6}>
+					<Form.Group>
+						<Form.Control as='select'>
+							<option>7 Add-Ons</option>
+						</Form.Control>
+					</Form.Group>
+				</Col>
+				<Col md={6}>
+					<Form.Group>
+						<Form.Control type='text' placeholder='' defaultValue='$15' />
```
Comment (sahilvicesoftware) [inline webapp/src/features/UserItemModals/AddInvoiceModal.js]: use `defaultValue='$15'` from a const

File: webapp/src/features/UserItem/userItem.selectors.js [hunk 23]
```diff
@@ -0,0 +1,6 @@
+import slice from './userItem.slice'
+
+export const selectSlice = (state) => state[slice.name]
+
+export const selectedUserItemModal = (state) =>
```
Comment (sahilvicesoftware) [inline webapp/src/features/UserItem/userItem.selectors.js]: selectUserItemModal

File: webapp/src/features/UserItem/BillingHistory.js (hunk 4 above)
Comment (sahilvicesoftware) [inline webapp/src/features/UserItem/BillingHistory.js:19]: please add a TODO here that we need to fix this

File: webapp/src/features/UserItem/ActivityHistory.js [hunk 24]
```diff
@@ -0,0 +1,138 @@
...
+import Table from '../../widgets/Table'
+import classNames from 'classnames'
+import SortingCustomIcon from '../../widgets/SortingCustomIcon'
+
+const ActivityHistory = () => {
+	const eventIconFormatter = (row) => {
+		let iconName = ''
+		let iconClass = ''
+		switch (true) {
+			case row.includes('Credit'):
```
Comment (sahilvicesoftware) [inline webapp/src/features/UserItem/ActivityHistory.js]: use enums for event types

**Review by sahilvicesoftware — CHANGES_REQUESTED** [review]

File: webapp/src/Routes.js (hunk 11 above)
Comment (ghost) [reply webapp/src/Routes.js:3]: updated

File: webapp/src/Routes.js (hunk 12 above)
Comment (ghost) [reply webapp/src/Routes.js]: updated

File: webapp/src/app.css (hunk 13 above)
Comment (ghost) [reply webapp/src/app.css:2]: this is same color code from spec

File: webapp/src/app.css (hunk 14 above)
Comment (ghost) [reply webapp/src/app.css:253]: updated

File: webapp/src/widgets/NavBar/NavBar.js (hunk 16 above)
Comment (ghost) [reply webapp/src/widgets/NavBar/NavBar.js]: updated the condition. This is not in progress.

File: webapp/src/widgets/Modal/Modal.js (hunk 17 above)
Comment (ghost) [reply webapp/src/widgets/Modal/Modal.js:6]: All the modal style has been moved to app.css due to size of the modal being incorrectly displayed while keeping it in Modal folder.

File: webapp/src/features/Dashboard/SubscriptionTable.js (hunk 18 above)
Comment (ghost) [reply webapp/src/features/Dashboard/SubscriptionTable.js:23]: updated

File: webapp/src/widgets/Cards/Cards.js (hunk 20 above)
Comment (ghost) [reply webapp/src/widgets/Cards/Cards.js]: updated

File: webapp/src/features/UserItemModals/AddInvoiceModal.js (hunk 22 above)
Comment (ghost) [reply webapp/src/features/UserItemModals/AddInvoiceModal.js]: updated

File: webapp/src/features/UserItem/userItem.selectors.js (hunk 23 above)
Comment (ghost) [reply webapp/src/features/UserItem/userItem.selectors.js]: updated

File: webapp/src/features/UserItem/BillingHistory.js (hunk 4 above)
Comment (ghost) [reply webapp/src/features/UserItem/BillingHistory.js:19]: updated

File: webapp/src/widgets/NavBar/NavBar.js (hunk 15 above)
Comment (ghost) [reply webapp/src/widgets/NavBar/NavBar.js:4]: updated

File: webapp/src/features/UserItem/ActivityHistory.js (hunk 24 above)
Comment (ghost) [reply webapp/src/features/UserItem/ActivityHistory.js]: updated

File: webapp/src/widgets/Cards/Cards.js (hunk 21 above)
Comment (ghost) [reply webapp/src/widgets/Cards/Cards.js]: updated as discussed
